Building from source fails on Fedora 40

### Bambu Studio Version

v01.09.01.58

### Where is the application from?

Built from source by yourself or others

### OS version

Fedora 40, Linux 6.8.8-300.fc40.x86_64 #1 SMP PREEMPT_DYNAMIC Sat Apr 27 17:53:31 UTC 2024 x86_64 GNU/Linux

### Additional system information

gcc version 14.0.1 20240411 (Red Hat 14.0.1-0) (GCC)
cmake version 3.28.2
GNU Make 4.4.1 Built for x86_64-redhat-linux-gnu

### Printer

Bambu Lab X1C

### How to reproduce

1. Clone this repository on Fedora 40 with the latest system updates
2. Follow the steps in the [compile guide](https://github.com/bambulab/BambuStudio/wiki/Linux-Compile-Guide)

### Actual results

After executing the `cmake --build . --target install --config Release -j1` command, I get the following compile-time error:

```
[  3%] Built target clipper
[  3%] Building CXX object src/clipper2/CMakeFiles/Clipper2.dir/Clipper2Lib/src/clipper.engine.cpp.o
In file included from /home/alok/src/BambuStudio/src/clipper2/Clipper2Lib/include/clipper2/clipper.engine.h:22,
                 from /home/alok/src/BambuStudio/src/clipper2/Clipper2Lib/src/clipper.engine.cpp:16:
/home/alok/src/BambuStudio/src/clipper2/Clipper2Lib/include/clipper2/clipper.core.h:117:26: error: template-id not allowed for constructor in C++20 [-Werror=template-id-cdtor]
  117 |         explicit Point<T>(const Point<T2>& p) { Init(p.x, p.y); }
      |                          ^
/home/alok/src/BambuStudio/src/clipper2/Clipper2Lib/include/clipper2/clipper.core.h:117:26: note: remove the ‘< >’
cc1plus: all warnings being treated as errors
gmake[2]: *** [src/clipper2/CMakeFiles/Clipper2.dir/build.make:76: src/clipper2/CMakeFiles/Clipper2.dir/Clipper2Lib/src/clipper.engine.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:699: src/clipper2/CMakeFiles/Clipper2.dir/all] Error 2
gmake: *** [Makefile:166: all] Error 2
```

g++ is throwing a C++20 error, even though src/clipper2/CMakeLists.txt explicitly specifies C++17:

```
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)
```

### Expected results

Compilation should complete without errors

### Project file & Debug log uploads

N/A

### Checklist of files to include

- [X] Log file
- [X] Project file

I investigated further. CMake is doing the right thing - it adds a `-std=c++17` flag for the compiler. What's happening is that gcc detects the `template-id-cdtor` issue and throws a warning. However, clipper2's CMakeLists.txt file passes the `-Werror` flag to gcc, which causes this warning to turn into an error. 

Adding `-Wno-template-id-cdtor` flag to clipper2's `target_compile_options` gets past that error

However, now I'm running into another build issue (Only showing the error here for clarity. If you need the whole build log, please let me know.):

```
/home/alok/src/BambuStudio_dep/usr/local/include/boost-1_78/boost/lexical_cast/detail/converter_lexical.hpp:243:13: error: static assertion failed: Target type is neither std::istream`able nor std::wistream`able
  243 |             BOOST_STATIC_ASSERT_MSG((result_t::value || boost::has_right_shift<std::basic_istream<wchar_t>, T >::value),
      |             ^~~~~~~~~~~~~~~~~~~~~~~
/home/alok/src/BambuStudio_dep/usr/local/include/boost-1_78/boost/lexical_cast/detail/converter_lexical.hpp:243:13: note: ‘(((bool)boost::integral_constant<bool, false>::value) || ((bool)boost::integral_constant<bool, false>::value))’ evaluates to false
gmake[2]: *** [src/libslic3r/CMakeFiles/libslic3r_cgal.dir/build.make:76: src/libslic3r/CMakeFiles/libslic3r_cgal.dir/MeshBoolean.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:994: src/libslic3r/CMakeFiles/libslic3r_cgal.dir/all] Error 2
gmake: *** [Makefile:166: all] Error 2
```

@lanewei120 Unfortunately, that didn't fix the issue. This is because the error is unrelated to that file. g++ is being stupid and carrying over the filename and line number from a previous warning (it is shown just before the top line in your screenshot):

```
/home/alok/src/BambuStudio/src/libslic3r/MeshBoolean.cpp: In function ‘bool Slic3r::MeshBoolean::mcut::do_boolean_single(McutMesh&, const McutMesh&, const std::string&)’:
/home/alok/src/BambuStudio/src/libslic3r/MeshBoolean.cpp:695:23: warning: comparison of integer expressions of different signedness: ‘int’ and ‘uint32_t’ {aka ‘unsigned int’} [-Wsign-compare]
  695 |     for (int n = 0; n < numConnComps; ++n) {
      |                     ~~^~~~~~~~~~~~~~ 
```

... so making the change you suggested removes that warning, but it doesn't fix the error.

> I can not reproduce it on my fedora 37, ubuntu 22.04
Yup, I looked at the CI build log on github and it seems to be fine as well. I think this is an issue that started to show up because of some build tool upgrade after fedora 37. Maybe gcc or cmake? What is the version of cmake and gcc on your setup?

--- a/src/clipper2/CMakeLists.txt
+++ b/src/clipper2/CMakeLists.txt
@@ -37,7 +37,7 @@ target_include_directories(Clipper2
 if (WIN32)
   target_compile_options(Clipper2 PRIVATE /W4 /WX)
 else()
-  target_compile_options(Clipper2 PRIVATE -Wall -Wextra -Wpedantic -Werror)
+  target_compile_options(Clipper2 PRIVATE -Wall -Wextra -Wpedantic -Werror -Wno-error=template-id-cdtor)
   target_link_libraries(Clipper2 PUBLIC -lm)
 
 endif()

--- a/src/clipper2/CMakeLists.txt
+++ b/src/clipper2/CMakeLists.txt
@@ -37,9 +37,11 @@ target_include_directories(Clipper2
 if (WIN32)
   target_compile_options(Clipper2 PRIVATE /W4 /WX)
 else()
-  target_compile_options(Clipper2 PRIVATE -Wall -Wextra -Wpedantic -Werror -Wno-error=template-id-cdtor)
+  target_compile_options(Clipper2 PRIVATE -Wall -Wextra -Wpedantic -Werror)
   target_link_libraries(Clipper2 PUBLIC -lm)
-
+  if (CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 14.1)
+    target_compile_options(Clipper2 PRIVATE -Wno-error=template-id-cdtor)
+  endif()
 endif()
 
 set_target_properties(Clipper2 PROPERTIES FOLDER Libraries

--- a/src/libslic3r/CMakeLists.txt
+++ b/src/libslic3r/CMakeLists.txt
@@ -456,6 +456,11 @@ if (_opts)
     target_compile_options(libslic3r_cgal PRIVATE "${_opts_bad}")
 endif()
 
+set_property(TARGET libslic3r PROPERTY CXX_STANDARD 17)
+set_property(TARGET libslic3r PROPERTY CXX_EXTENSIONS OFF)
+set_property(TARGET libslic3r_cgal PROPERTY CXX_STANDARD 17)
+set_property(TARGET libslic3r_cgal PROPERTY CXX_EXTENSIONS OFF)
+
 target_link_libraries(libslic3r_cgal PRIVATE ${_cgal_tgt} libigl mcut)
 
 if (MSVC AND "${CMAKE_SIZEOF_VOID_P}" STREQUAL "4") # 32 bit MSVC workaround

--- a/src/libslic3r/CMakeLists.txt
+++ b/src/libslic3r/CMakeLists.txt
@@ -456,10 +456,12 @@ if (_opts)
     target_compile_options(libslic3r_cgal PRIVATE "${_opts_bad}")
 endif()
 
-set_property(TARGET libslic3r PROPERTY CXX_STANDARD 17)
-set_property(TARGET libslic3r PROPERTY CXX_EXTENSIONS OFF)
-set_property(TARGET libslic3r_cgal PROPERTY CXX_STANDARD 17)
-set_property(TARGET libslic3r_cgal PROPERTY CXX_EXTENSIONS OFF)
+if (CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 14.1)
+    set_property(TARGET libslic3r PROPERTY CXX_STANDARD 17)
+    set_property(TARGET libslic3r PROPERTY CXX_EXTENSIONS OFF)
+    set_property(TARGET libslic3r_cgal PROPERTY CXX_STANDARD 17)
+    set_property(TARGET libslic3r_cgal PROPERTY CXX_EXTENSIONS OFF)
+endif()
 
 target_link_libraries(libslic3r_cgal PRIVATE ${_cgal_tgt} libigl mcut)
 

